File: backend/app/schemas/company_library.py
"""Company Library (MDM) — Phase 1 Pydantic v2 schemas."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ── Field definitions ──────────────────────────────────────────────────

FieldType = Literal["number", "text", "date", "enum", "formula", "boolean"]
ScopeType = Literal["all", "sector", "companies"]
Layout    = Literal["one_col", "two_col", "grid"]


class FieldDefinitionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    code: str
    name_ru: str
    name_uz: Optional[str] = None
    name_en: Optional[str] = None
    field_type: FieldType
    unit: Optional[str] = None
    format_pattern: Optional[str] = None
    enum_values: Optional[Any] = None
    formula: Optional[str] = None
    scope_type: ScopeType
    scope_value: Optional[Any] = None
    source_module: Optional[str] = None
    source_path: Optional[str] = None
    permission_view: Optional[str] = None
    permission_edit: Optional[str] = None
    is_system: bool
    sort_order: int
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime


class FieldDefinitionCreate(BaseModel):
    code: str = Field(..., min_length=1, max_length=128, pattern=r"^[a-z][a-z0-9_]*$")
    name_ru: str = Field(..., min_length=1, max_length=255)
    name_uz: Optional[str] = Field(None, max_length=255)
    name_en: Optional[str] = Field(None, max_length=255)
    field_type: FieldType
    unit: Optional[str] = Field(None, max_length=32)
    format_pattern: Optional[str] = Field(None, max_length=64)
    enum_values: Optional[list[str]] = None
    formula: Optional[str] = None
    scope_type: ScopeType = "all"
    scope_value: Optional[Any] = None
    source_module: Optional[str] = Field(None, max_length=64)
    source_path: Optional[str] = Field(None, max_length=255)
    permission_view: Optional[str] = Field(None, max_length=128)
    permission_edit: Optional[str] = Field(None, max_length=128)
    sort_order: int = 100


class FieldDefinitionUpdate(BaseModel):
    name_ru: Optional[str] = Field(None, min_length=1, max_length=255)
    name_uz: Optional[str] = Field(None, max_length=255)
    name_en: Optional[str] = Field(None, max_length=255)
    unit: Optional[str] = Field(None, max_length=32)
    format_pattern: Optional[str] = Field(None, max_length=64)
    enum_values: Optional[list[str]] = None
    formula: Optional[str] = None
    scope_type: Optional[ScopeType] = None
    scope_value: Optional[Any] = None
    permission_view: Optional[str] = Field(None, max_length=128)
    permission_edit: Optional[str] = Field(None, max_length=128)
    sort_order: Optional[int] = None


# ── Library views (per-user column prefs) ──────────────────────────────

class LibraryViewRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    name: str
    is_default: bool
    visible_columns: list[str]
    filters: dict[str, Any]
    sort_by: Optional[str] = None
    sort_dir: str
    created_at: datetime
    updated_at: datetime


class LibraryViewCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    is_default: bool = False
    visible_columns: list[str] = Field(default_factory=list)
    filters: dict[str, Any] = Field(default_factory=dict)
    sort_by: Optional[str] = Field(None, max_length=64)
    sort_dir: Literal["asc", "desc"] = "desc"


class LibraryViewUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=128)
    is_default: Optional[bool] = None
    visible_columns: Optional[list[str]] = None
    filters: Optional[dict[str, Any]] = None
    sort_by: Optional[str] = Field(None, max_length=64)
    sort_dir: Optional[Literal["asc", "desc"]] = None


# ── Library tabs ───────────────────────────────────────────────────────

class LibraryTabRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    code: str
    name_ru: str
    name_uz: Optional[str] = None
    name_en: Optional[str] = None
    field_codes: list[str]
    layout: Layout
    is_system: bool
    sort_order: int
    scope_type: ScopeType
    scope_value: Optional[Any] = None
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime


class LibraryTabCreate(BaseModel):
    code: str = Field(..., min_length=1, max_length=128, pattern=r"^[a-z][a-z0-9_]*$")
    name_ru: str = Field(..., min_length=1, max_length=255)
    name_uz: Optional[str] = Field(None, max_length=255)
    name_en: Optional[str] = Field(None, max_length=255)
    field_codes: list[str] = Field(default_factory=list)
    layout: Layout = "two_col"
    sort_order: int = 100
    scope_type: ScopeType = "all"
    scope_value: Optional[Any] = None


class LibraryTabUpdate(BaseModel):
    name_ru: Optional[str] = Field(None, min_length=1, max_length=255)
    name_uz: Optional[str] = Field(None, max_length=255)
    name_en: Optional[str] = Field(None, max_length=255)
    field_codes: Optional[list[str]] = None
    layout: Optional[Layout] = None
    sort_order: Optional[int] = None
    scope_type: Optional[ScopeType] = None
    scope_value: Optional[Any] = None


# ── Library list/detail responses ──────────────────────────────────────

class LibraryCompanyRow(BaseModel):
    """Row in the library index. `fields` is a flat dict keyed by field code."""
    id: UUID
    code: Optional[str] = None
    name_ru: str
    name_short: Optional[str] = None
    sector_id: Optional[UUID] = None
    sector_name: Optional[str] = None
    fields: dict[str, Any] = Field(default_factory=dict)


class LibraryListResponse(BaseModel):
    items: list[LibraryCompanyRow]
    total: int
    columns: list[FieldDefinitionRead]
    available_views: list[LibraryViewRead] = Field(default_factory=list)
    active_view_id: Optional[UUID] = None


class LibraryFieldValue(BaseModel):
    """One field's value on a detail-view, with provenance."""
    code: str
    value: Any = None
    source_module: Optional[str] = None
    source_updated_at: Optional[datetime] = None
    source_actor: Optional[str] = None


class LibraryActivityEntry(BaseModel):
    ts: datetime
    actor_email: Optional[str] = None
    module: Optional[str] = None
    action: str
    field_code: Optional[str] = None
    diff: Optional[dict[str, Any]] = None


class LibraryCompanyDetail(BaseModel):
    company_id: UUID
    company_code: Optional[str] = None
    company_name: str
    sector_id: Optional[UUID] = None
    sector_name: Optional[str] = None
    fields: list[LibraryFieldValue]
    tabs: list[LibraryTabRead]
    activity: list[LibraryActivityEntry] = Field(default_factory=list)


# ── Field-write payload ────────────────────────────────────────────────

class FieldWriteRequest(BaseModel):
    value: Any = None
    reason: Optional[str] = None


class FieldWriteResponse(BaseModel):
    code: str
    value: Any = None
    source_module: Optional[str] = None
    updated_at: datetime
    routed_to: Optional[str] = None  # textual hint where the write went

File: frontend/src/api/companyLibrary.ts
/**
 * Company Library (MDM) API client — Phase 2 frontend.
 * Mirrors backend/app/api/routes/company_library.py + schemas/company_library.py.
 */
import { api } from "./client";

// ── Types ─────────────────────────────────────────────────────────────

export type FieldType   = "number" | "text" | "date" | "enum" | "formula" | "boolean";
export type ScopeType   = "all"    | "sector" | "companies";
export type TabLayout   = "one_col"| "two_col" | "grid";

export interface FieldDefinition {
  id: string;
  code: string;
  name_ru: string;
  name_uz: string | null;
  name_en: string | null;
  field_type: FieldType;
  unit: string | null;
  format_pattern: string | null;
  enum_values: string[] | null;
  formula: string | null;
  scope_type: ScopeType;
  scope_value: any;
  source_module: string | null;
  source_path: string | null;
  permission_view: string | null;
  permission_edit: string | null;
  is_system: boolean;
  sort_order: number;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface FieldDefinitionCreate {
  code: string;
  name_ru: string;
  name_uz?: string | null;
  name_en?: string | null;
  field_type: FieldType;
  unit?: string | null;
  format_pattern?: string | null;
  enum_values?: string[] | null;
  formula?: string | null;
  scope_type?: ScopeType;
  scope_value?: any;
  source_module?: string | null;
  source_path?: string | null;
  permission_view?: string | null;
  permission_edit?: string | null;
  sort_order?: number;
}

export interface LibraryView {
  id: string;
  user_id: string;
  name: string;
  is_default: boolean;
  visible_columns: string[];
  filters: Record<string, any>;
  sort_by: string | null;
  sort_dir: "asc" | "desc";
  created_at: string;
  updated_at: string;
}

export interface LibraryTab {
  id: string;
  code: string;
  name_ru: string;
  name_uz: string | null;
  name_en: string | null;
  field_codes: string[];
  layout: TabLayout;
  is_system: boolean;
  sort_order: number;
  scope_type: ScopeType;
  scope_value: any;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface LibraryCompanyRow {
  id: string;
  code: string | null;
  name_ru: string;
  name_short: string | null;
  sector_id: string | null;
  sector_name: string | null;
  fields: Record<string, any>;
}

export interface LibraryListResponse {
  items: LibraryCompanyRow[];
  total: number;
  columns: FieldDefinition[];
  available_views: LibraryView[];
  active_view_id: string | null;
}

export interface LibraryFieldValue {
  code: string;
  value: any;
  source_module: string | null;
  source_updated_at: string | null;
  source_actor: string | null;
}

export interface LibraryActivityEntry {
  ts: string;
  actor_email: string | null;
  module: string | null;
  action: string;
  field_code: string | null;
  diff: Record<string, any> | null;
}

export interface LibraryCompanyDetail {
  company_id: string;
  company_code: string | null;
  company_name: string;
  sector_id: string | null;
  sector_name: string | null;
  fields: LibraryFieldValue[];
  tabs: LibraryTab[];
  activity: LibraryActivityEntry[];
}

// ── Endpoints ─────────────────────────────────────────────────────────

export const companyLibraryApi = {
  list(params: {
    sector?: string;
    search?: string;
    view_id?: string;
    limit?: number;
    offset?: number;
  } = {}) {
    return api.get<LibraryListResponse>("/library/companies", { params }).then(r => r.data);
  },

  detail(companyId: string) {
    return api.get<LibraryCompanyDetail>(`/library/companies/${companyId}`).then(r => r.data);
  },

  updateField(companyId: string, code: string, value: any, reason?: string) {
    return api.patch(`/library/companies/${companyId}/fields/${code}`, {
      value, reason: reason ?? null,
    }).then(r => r.data);
  },

  // 30-сек тикет для sync-WebSocket (уходит в субпротокол, не в URL).
  wsTicket() {
    return api.post<{ ticket: string; expires_in: number }>(
      "/companies/ws-ticket",
    ).then(r => r.data);
  },

  activity(companyId: string, limit = 10) {
    return api.get<LibraryActivityEntry[]>(
      `/library/companies/${companyId}/activity`, { params: { limit } },
    ).then(r => r.data);
  },

  // ── Field definitions ──
  listFields(params: { sector?: string; scope_type?: string } = {}) {
    return api.get<FieldDefinition[]>("/field-definitions", { params }).then(r => r.data);
  },
  createField(payload: FieldDefinitionCreate) {
    return api.post<FieldDefinition>("/field-definitions", payload).then(r => r.data);
  },
  updateField_def(code: string, patch: Partial<FieldDefinitionCreate>) {
    return api.patch<FieldDefinition>(`/field-definitions/${code}`, patch).then(r => r.data);
  },
  deleteField_def(code: string) {
    return api.delete(`/field-definitions/${code}`);
  },

  // ── Views ──
  listViews() {
    return api.get<LibraryView[]>("/library-views").then(r => r.data);
  },
  createView(payload: { name: string; is_default?: boolean; visible_columns?: string[];
                        filters?: Record<string, any>; sort_by?: string; sort_dir?: "asc" | "desc" }) {
    return api.post<LibraryView>("/library-views", payload).then(r => r.data);
  },
  updateView(id: string, patch: Partial<LibraryView>) {
    return api.patch<LibraryView>(`/library-views/${id}`, patch).then(r => r.data);
  },
  deleteView(id: string) {
    return api.delete(`/library-views/${id}`);
  },

  // ── Tabs ──
  listTabs() {
    return api.get<LibraryTab[]>("/library-tabs").then(r => r.data);
  },
  createTab(payload: {
    code: string; name_ru: string; field_codes?: string[]; layout?: TabLayout;
    sort_order?: number; scope_type?: ScopeType; scope_value?: any;
  }) {
    return api.post<LibraryTab>("/library-tabs", payload).then(r => r.data);
  },
  updateTab(code: string, patch: Partial<LibraryTab>) {
    return api.patch<LibraryTab>(`/library-tabs/${code}`, patch).then(r => r.data);
  },
  deleteTab(code: string) {
    return api.delete(`/library-tabs/${code}`);
  },
};

// ── WebSocket envelope ────────────────────────────────────────────────

export interface FieldUpdateEvent {
  type: "field_update";
  company_id: string;
  field_code: string;
  value: any;
  source_module: string | null;
  actor_id: string | null;
  ts: number;
}
